packages/flutter/lib/src/material/input_decorator.dart

```diff
@@ -2260,20 +2262,24 @@ class _InputDecoratorState extends State<InputDecorator> with TickerProviderStat
         textAlign: textAlign,
         maxLines: decoration.hintMaxLines,
       );
-      hint =
-          maintainHintHeight
-              ? AnimatedOpacity(
-                opacity: showHint ? 1.0 : 0.0,
-                duration: decoration.hintFadeDuration ?? _kHintFadeTransitionDuration,
-                curve: _kTransitionCurve,
-                child: hintTextWidget,
-              )
-              : AnimatedSwitcher(
-                duration: decoration.hintFadeDuration ?? _kHintFadeTransitionDuration,
-                transitionBuilder: _buildTransition,
-                child: showHint ? hintTextWidget : const SizedBox.shrink(),
-              );
+    } else {
+      hintTextWidget = const SizedBox.shrink();
     }
+    final bool showHint = isEmpty && !_hasInlineLabel;
+    hint =
+        maintainHintHeight
+            ? AnimatedOpacity(
+              opacity: showHint ? 1.0 : 0.0,
+              duration: decoration.hintFadeDuration ?? _kHintFadeTransitionDuration,
+              curve: _kTransitionCurve,
+              child: hintTextWidget,
+            )
+            : AnimatedSwitcher(
+              duration: decoration.hintFadeDuration ?? _kHintFadeTransitionDuration,
+              transitionBuilder: _buildTransition,
+              child: showHint ? hintTextWidget : const SizedBox.shrink(),
+            );
+
```

Comment:
This code will always add a hint widget. Previously, hint was null when hintText was null.
This will break logic that might expect hint widget to be null when not defined.
It is safer to keep the change as narrow as needed: add a non-null hint only when (widget.hintText != null || widget.hint != null).

Reply:
Kept the changes minimal and added a null check to leave `hintTextWidget` null when `hint` & `hintText` is null